File: apps/server/api/tags/tags.go
package tags

import (
	"booru-viewer/server/api"
	"booru-viewer/server/api/model"
	"encoding/json"
	"fmt"
	"net/http"
)

type Order string

const (
	Asc		Order = "asc"
	Desc	Order = "desc"
)

type OrderBy string

const (
	Date	OrderBy = "date"
	Count	OrderBy = "count"
	Name	OrderBy = "name"
)

func FetchTags(c *api.Client, limit int, page int, search string, order Order, orderby OrderBy, tagID *int) (model.Tags, *api.ClientError) {
	var url string
	if tagID != nil {
		url = fmt.Sprintf("%s&s=tag&q=index&json=1&id=%d", c.BaseAPI(), *tagID);
	} else {
		url = fmt.Sprintf("%s&s=tag&q=index&json=1&limit=%d&pid=%d&name_pattern=%s%%&order=%s&orderby=%s", c.BaseAPI(), limit, page - 1, search, order, orderby);
	}

	resp, err := c.HTTPClient().Get(url);
	if err != nil {
		return model.Tags{}, &api.ClientError{
			StatusCode: http.StatusServiceUnavailable,
			Message: fmt.Sprintf("request error: %s", err),
		}
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return model.Tags{}, &api.ClientError{
			StatusCode: resp.StatusCode,
			Message: fmt.Sprintf("unexpected status code: %d", resp.StatusCode),
		}
	}

	var tags model.Tags
	if err := json.NewDecoder(resp.Body).Decode(&tags); err != nil {
		return model.Tags{}, &api.ClientError{
			StatusCode: resp.StatusCode,
			Message: fmt.Sprintf("decoding error: %s", err),
		}
	}

	return tags, nil
}

File: packages/api-client/src/tags/tags.ts
import { FetchSource, TagsResponse } from "../typings";
import { apiClient } from "../client";
import { AxiosResponse } from "axios";

export type Order = "date" | "count" | "name";

export const tagsApi = {
  getAllTags: async (
    source: FetchSource,
    opts?: {
      page: number;
      limit: number;
      search?: string;
      order?: Order;
    }
  ): Promise<AxiosResponse<TagsResponse>> => {
    const params = opts
      ? {
          page: opts.page,
          limit: opts.limit,
          search: opts.search || "",
          order: opts.order || "count",
        }
      : {};

    return apiClient.get(`/${source}/tags`, {
      params,
    });
  },

  getSingleTag: async (
    source: FetchSource,
    tagID: number
  ): Promise<AxiosResponse<TagsResponse>> => {
    return apiClient.get(`/${source}/posts/${tagID}`);
  },
};
